### PR TITLE
Update: add extra phase sifters, support autograd

### PR DESCRIPTION
**Pull request by royess**

### CODE REVIEW

File: circuits_simulator.py [hunk 1]
```diff
@@ -8,19 +8,20 @@
...
 '''
 def mzi(alpha, phi):
-    return np.array([
-        [np.exp(1j*phi) * np.sin(alpha/2), np.exp(1j*phi) * np.cos(alpha/2)],
-        [np.cos(alpha/2), -np.sin(alpha/2)]
-    ]) # TODO: whether this is a good expression!
+    return (1/2) * np.array([
+        [np.exp(1j*phi) * (np.exp(1j*alpha)+1), np.exp(1j*alpha)-1],
+        [np.exp(1j*phi) * (np.exp(1j*alpha)-1), np.exp(1j*alpha)+1]
+    ]) # TODO: whether this is the right expression!
```
Comment (michaelchoquer) [inline circuits_simulator.py:13]: Not seeing wherre you're getting this.  I think going back to the original (unitary) mzi definition makes sense.

File: circuits_simulator.py [hunk 2]
```diff
@@ -38,7 +39,8 @@ def __init__(self, numModes, alphaArr, phiArr):
                 np.eye(pair[1]-1),
                 mzi(alphaArr[pair[0]-1,pair[1]-1], phiArr[pair[0]-1,pair[1]-1]),
                 np.eye(numModes-1-pair[1]))
```
Comment (michaelchoquer) [inline circuits_simulator.py]: Misssing where the +/- 1's are coming from.

**Review by michaelchoquer — CHANGES_REQUESTED** [review]
Not sure about mzi(): can you show me the reference(s) where you got that?

**Comment by royess** [comment]
Actually I'm not very confident about how mzi should be implemented all the time... And original mzi implement is more like a guess. So in this PR, I refer to the implement from bosonic package, although it makes previous parameters fail to produce CNOT gate.

---

Bosonic mzi link: https://github.com/steinbrecher/bosonic/blob/master/bosonic/reck.py

**Pull request merged**